debezium-embedded/src/main/java/io/debezium/embedded/ConnectorEngine.java

```diff
@@ -0,0 +1,1548 @@
+                offsetCommitPolicy = OffsetCommitPolicy.periodic(offsetPeriodMs, TimeUnit.MILLISECONDS);
+
+                // Start any existing connector workers ...
+                workers.forEach((connectorName, worker) -> {
+                    logger.debug("Starting connector '{}' in engine '{}'", connectorName, name());
+                    worker.start();
+                });
+            } catch (Throwable t) {
+                running.set(false);
+            }
+        }
+    }
+
+    /**
+     * Wrap the supplied offset storage reader with a delegate implementation that also logs the partitions and offsets that are
+     * read.
+     * 
+     * @param reader the actual offset reader; may not be null
+     * @param logger the logger to which trace messages should be written; may not be null
+     * @return the instrumented reader implementation; never null
+     */
+    protected static OffsetStorageReader instrument(OffsetStorageReader reader, Logger logger) {
+        return new OffsetStorageReader() {
+            @Override
+            public <T> Map<Map<String, T>, Map<String, Object>> offsets(Collection<Map<String, T>> partitions) {
+                logger.trace("Reading offsets for {} partitions", partitions.size());
+                Map<Map<String, T>, Map<String, Object>> results = reader.offsets(partitions);
+                if (logger.isTraceEnabled()) {
+                    logger.trace("Read offsets for {} partitions:", partitions.size());
+                    AtomicInteger partitionNumber = new AtomicInteger();
+                    results.forEach((partition, offset) -> {
+                        logger.trace("Offset for partition {} {}: {}", partitionNumber.incrementAndGet(), partition, offset);
+                    });
+                }
+                return results;
+            }
+
+            @Override
+            public <T> Map<String, Object> offset(Map<String, T> partition) {
+                Map<String, Object> offset = reader.offset(partition);
+                logger.trace("Offset for partition {}: {}", partition, offset);
+                return offset;
+            }
+        };
+    }
+
+    /**
+     * Determine if this engine has been {@link #start() started} and is still running.
+     * 
+     * @return {@code true} if this engine is currently running, or {@code false} otherwise
+     */
+    public boolean isRunning() {
+        return running.get();
+    }
+
+    @Override
+    public List<ChangeEvent> poll() throws InterruptedException {
+        if (shutdown.get()) {
+            return pollResultsWhenShutdown();
+        }
+        return withEvents(eventQueue.remove());
+    }
+
+    @Override
+    public List<ChangeEvent> poll(long timeout, TimeUnit unit) throws InterruptedException {
+        if (shutdown.get()) {
+            return pollResultsWhenShutdown();
+        }
+        return withEvents(eventQueue.poll(timeout, unit));
```

Comment:
I'm not convinced using `poll` is a good idea because if  I understand the intent behind this class, it's supposed to run concurrently multiple tasks from multiple connectors. So in a real system, the back-pressure on this queue will be significant and therefore IMO `drainTo` a certain batch makes more sense instead of `poll`ing one element at a time